### PR TITLE
8263560: Remove needless wrapping with BufferedInputStream

### PR DESCRIPTION
**Pull request by stsypanov**
In some cases wrapping of `InputStream` with `BufferedInputStream` is redundant, e.g. in case the wrapped one is `ByteArrayOutputStream` which does not require any buffer having one within.

Other cases are related to reading either a byte or short `byte[]`: in both cases `BufferedInputStream.fill()` will be called resulting in load of much bigger amount of data (8192 by default) than required.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263560](https://bugs.openjdk.java.net/browse/JDK-8263560): Remove needless wrapping with BufferedInputStream

### Reviewers
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to a016d2ac59c073987e994e9053395b483a55f515
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to ff25cc4d6d19cb2465e2a15087294a25750f25fa
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2992/head:pull/2992`
`$ git checkout pull/2992`

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/jmod/JmodFile.java [hunk 1]
```diff
@@ -52,11 +51,10 @@
     };
 
     public static void checkMagic(Path file) throws IOException {
-        try (InputStream in = Files.newInputStream(file);
-             BufferedInputStream bis = new BufferedInputStream(in)) {
+        try (InputStream in = Files.newInputStream(file)) {
             // validate the header
             byte[] magic = new byte[4];
-            bis.read(magic);
+            in.read(magic);
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/jmod/JmodFile.java:57]: While you are there, this can be changed to magic = in.readNBytes(4) and throw IOException if magic.length != 4.

Comment (stsypanov) [reply src/java.base/share/classes/jdk/internal/jmod/JmodFile.java]: Done.

File: src/java.base/share/classes/jdk/internal/jmod/JmodFile.java [hunk 2]
```diff
@@ -52,11 +51,12 @@
...
     public static void checkMagic(Path file) throws IOException {
-        try (InputStream in = Files.newInputStream(file);
-             BufferedInputStream bis = new BufferedInputStream(in)) {
+        try (InputStream in = Files.newInputStream(file)) {
             // validate the header
-            byte[] magic = new byte[4];
-            bis.read(magic);
+            byte[] magic = in.readNBytes(4);
+            if (magic.length != 4) {
+                throw new IOException("Header expected to be of length 4, but was " + magic.length);
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/jmod/JmodFile.java:58]: Thanks for the update. If magic != 4 then it means the file is not a JMOD file or that it has been truncated.

Comment (stsypanov) [reply src/java.base/share/classes/jdk/internal/jmod/JmodFile.java:58]: It turned out, that with this latest update come of tier2_tools tests are failing, e.g. `JLinkNegativeTest`:
```
test JLinkNegativeTest.testMalformedJmod(): failure
java.lang.AssertionError: Output does not fit regexp: Error: java.io.IOException: Invalid JMOD file
	at tests.Result.assertFailure(Result.java:70)
	at JLinkNegativeTest.testMalformedJmod(JLinkNegativeTest.java:201)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:821)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1131)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:773)
	at org.testng.TestRunner.run(TestRunner.java:623)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1185)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1110)
	at org.testng.TestNG.run(TestNG.java:1018)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:94)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:298)
	at java.base/java.lang.Thread.run(Thread.java:831)
...
java.lang.module.FindException: java.io.IOException: Invalid JMOD file: /home/s.tsypanov/IdeaProjects/jdk-github/build/linux-x86_64-server-release/test-support/jtreg_test_jdk_tier2/scratch/3/jmods/hacked4.jmod
	at java.base/jdk.internal.module.ModulePath.scan(ModulePath.java:253)
	at java.base/jdk.internal.module.ModulePath.scanNextEntry(ModulePath.java:190)
	at java.base/jdk.internal.module.ModulePath.find(ModulePath.java:154)
	at java.base/java.lang.module.Resolver.findWithBeforeFinder(Resolver.java:842)
	at java.base/java.lang.module.Resolver.resolve(Resolver.java:119)
	at java.base/java.lang.module.Configuration.resolve(Configuration.java:421)
	at java.base/java.lang.module.Configuration.resolve(Configuration.java:255)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.limitFinder(JlinkTask.java:545)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.newModuleFinder(JlinkTask.java:466)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.initJlinkConfig(JlinkTask.java:374)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.run(JlinkTask.java:267)
	at jdk.jlink/jdk.tools.jlink.internal.Main.run(Main.java:54)
	at jdk.jlink/jdk.tools.jlink.internal.Main$JlinkToolProvider.run(Main.java:65)
	at tests.JImageGenerator$JLinkTask.call(JImageGenerator.java:715)
	at tests.Helper.generateDefaultImage(Helper.java:257)
	at tests.Helper.generateDefaultImage(Helper.java:244)
	at JLinkNegativeTest.testSectionsAreFiles(JLinkNegativeTest.java:307)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:821)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1131)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:773)
	at org.testng.TestRunner.run(TestRunner.java:623)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1185)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1110)
	at org.testng.TestNG.run(TestNG.java:1018)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:94)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:298)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.io.IOException: Invalid JMOD file: /home/s.tsypanov/IdeaProjects/jdk-github/build/linux-x86_64-server-release/test-support/jtreg_test_jdk_tier2/scratch/3/jmods/hacked4.jmod
	at java.base/jdk.internal.jmod.JmodFile.checkMagic(JmodFile.java:62)
	at java.base/jdk.internal.jmod.JmodFile.<init>(JmodFile.java:180)
	at java.base/jdk.internal.module.ModulePath.readJMod(ModulePath.java:392)
	at java.base/jdk.internal.module.ModulePath.readModule(ModulePath.java:343)
	at java.base/jdk.internal.module.ModulePath.scanDirectory(ModulePath.java:284)
	at java.base/jdk.internal.module.ModulePath.scan(ModulePath.java:232)
	... 44 more

java.lang.Exception: failures: 1
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:96)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:298)
	at java.base/java.lang.Thread.run(Thread.java:831)
```

So if you don't mind, I'll revert the latest commit

Comment (AlanBateman) [reply src/java.base/share/classes/jdk/internal/jmod/JmodFile.java]: JLinkNegativeTest depends on the exception message, it should be easy to update.

Comment (stsypanov) [reply src/java.base/share/classes/jdk/internal/jmod/JmodFile.java:58]: Done

**Review by prrace — APPROVED** [review]
2d change is fine.

File: src/java.desktop/share/classes/sun/awt/image/ByteArrayImageSource.java [hunk 3]
```diff
@@ -52,10 +51,7 @@ final boolean checkSecurity(Object context, boolean quiet) {
     }
 
     protected ImageDecoder getDecoder() {
-        InputStream is =
-            new BufferedInputStream(new ByteArrayInputStream(imagedata,
-                                                             imageoffset,
-                                                             imagelength));
+        InputStream is = new ByteArrayInputStream(imagedata, imageoffset, imagelength);
```
Comment (mrserb) [inline src/java.desktop/share/classes/sun/awt/image/ByteArrayImageSource.java:54]: This file use 80 chars per line code style.

Comment (stsypanov) [reply src/java.desktop/share/classes/sun/awt/image/ByteArrayImageSource.java]: Fixed.

File: src/java.base/share/classes/jdk/internal/jmod/JmodFile.java (hunk 2 above)
Comment (AlanBateman) [reply src/java.base/share/classes/jdk/internal/jmod/JmodFile.java:58]: > Done

I think I'd prefer if the exception message would say that the JMOD is invalid or that the JMOD file is truncated (because I don't think anyone seeing this exception will know anything about the header).

Comment (stsypanov) [reply src/java.base/share/classes/jdk/internal/jmod/JmodFile.java]: Fixed

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/share/classes/sun/net/www/http/HttpClient.java [hunk 4]
```diff
@@ -428,8 +428,7 @@ protected boolean available() {
             try {
                 old = serverSocket.getSoTimeout();
                 serverSocket.setSoTimeout(1);
-                BufferedInputStream tmpbuf =
-                        new BufferedInputStream(serverSocket.getInputStream());
+                InputStream tmpbuf = serverSocket.getInputStream();
                 int r = tmpbuf.read();
                 if (r == -1) {
                     logFinest("HttpClient.available(): " +
```
Comment (dfuch) [inline src/java.base/share/classes/sun/net/www/http/HttpClient.java:434]: There are some subtle things going on there. Using a BufferedInputStream ensure that all the bytes available on the socket will be read, up to the buffer capacity. Can you revert this change? I'd rather that this clean up be handled separately. I have logged https://bugs.openjdk.java.net/browse/JDK-8263599.

**Review by dfuch — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/sun/net/www/http/HttpClient.java (hunk 4 above)
Comment (stsypanov) [reply src/java.base/share/classes/sun/net/www/http/HttpClient.java:434]: Sure, I'll revert this.

**Review by dfuch — APPROVED** [review]
LGTM. Thanks!

**Review by mrserb — APPROVED** [review]
